docs/cookbook/extending-the-ui.md

```diff
@@ -0,0 +1,990 @@
+
+```javascript
+addGradeFieldType() {
+  self.apos.schema.addFieldType({
+    name: 'gradeField',
+    convert: self.convertInput,
+    vueComponent: 'InputGradeField'
+  });
+}
+```
+
+We are passing the `addFieldType()` method of the `@apostrophecms/schema` module an object with three properties. The first is the name that will be used within any schema utilizing this field, `gradeField`. The `convert` key points to the method that will be used to sanitize and store the data. Finally, the `vueComponent` key takes the file name for our Vue component located in the `modules/grade/ui/apos/components` folder.
+
+
+```javascript
+exposeSchema() {
+  const schema = [
+  👉🏻{
+    👉🏻name: 'midterm',
+      type: 'integer',
+      label: 'Midterm Test',
+      required: true
+    },
+    {
+      name: 'final',
+      type: 'integer',
+      label: 'Final Test',
+      required: true
+    }
+  ];
+  return schema;
+}
+```
+
+The next method, `exposeSchema()`, allows us to reuse the schema for our component in the convert method, as well as in our Vue component. 
+
+The schema should be passed as an *array* of objects. This is slightly different than how we typically format our schema as an object composed of named objects, so the name field moves inside the object.
+
+```javascript
+async convertInput(req, field, data, object) {
+  data = data[field.name];
+  const schema = self.exposeSchema();
+  const errors = [];
+  const result = {
+    _id:
+      self.apos.launder.id(data && data._id) ||
+      self.apos.util.generateId()
+  };
+  if (data == null || typeof data !== 'object' || Array.isArray(data)) {
+    data = {};
+  }
+  ...
+}
+```
+The final method, `convertInput()`, accepts the data from the component inputs, makes sure the data is in an expected form, and then passes each data field through a `launder` method. The final data is returned as an object that gets stored by Apostrophe in the database. The first section of this method prepares the data for sanitization and storage.
+
+This method takes four arguments from the Vue component. The two most important for this recipe are the `data` and `object` arguments. `data` is going to contain all of the data being delivered from the input fields. We are only interested in managing the data from our new custom field, so in the very first line of the method we reassign `data` as being equal to only the data coming from our field using `data = data[field.name]`.
+
+We next pull in the schema we defined above using the `exposeSchema()` method. This schema is used to inform the `convert()` method of what types of data laundering methods to use.
+
+The `errors` and `results` variables are next initialized so that they can take values returned from the `@apostrophecms/schema` module `convert()` method.
+
+Note that for `result` we are making sure that an `_id` property exists and is sanitized, otherwise, we generate one.
```

Comment:
Remove the _id property, it is not necessary and probably isn't adding anything.

Reply:
Removed.